src/Processors/QueryPlan/ReadFromLoopStep.cpp

```diff
@@ -74,14 +76,24 @@ namespace DB
                     loop = true;
                 }
                 Chunk chunk;
+
+                if (query_info.trivial_limit > 0 && rows_read >= query_info.trivial_limit)
+                    return chunk;
+
                 if (executor && executor->pull(chunk))
                 {
-                    if (chunk)
-                    {
-                        retries_count = 0;
+                    rows_read += chunk.getNumRows();
+                    retries_count = 0;
+                    if (query_info.trivial_limit == 0 || rows_read <= query_info.trivial_limit)
                         return chunk;
-                    }
 
+                    size_t remaining_rows = query_info.trivial_limit + chunk.getNumRows() - rows_read;
+                    auto columns = chunk.detachColumns();
+                    for (auto & col : columns)
+                    {
+                        col = col->cut(0, remaining_rows);
+                    }
+                    return {std::move(columns), remaining_rows};
```

Comment:
Why do we need it? Won't `LIMIT` be applied later in `LimitTransform`?

Reply:
Without the change, LIMIT does not apply, and the following query runs in an infinite loop until cancelled:
```
SELECT * FROM loop(remote('localhost:9000', currentDatabase(), 't0')) tx LIMIT 3;
```
I'll take a look at the  LimitTransform, maybe something is missing.